src/main/java/gregtech/common/metatileentities/electric/MetaTileEntityMiner.java

```diff
@@ -0,0 +1,375 @@
+                    builder.widget(new SlotWidget(exportItems, index, 143 - rowSize * 9 + x * 18, 18 + y * 18, true, false)
+                            .setBackgroundTexture(GuiTextures.SLOT));
+                }
+            }
+        }
+
+        builder.image(7, 16, 105, 75, GuiTextures.DISPLAY)
+                .label(10, 5, getMetaFullName());
+        builder.widget(new AdvancedTextWidget(10, 23, this::addDisplayText, 0xFFFFFF)
+                .setMaxWidthLimit(84));
+        builder.widget(new AdvancedTextWidget(70, 23, this::addDisplayText2, 0xFFFFFF)
+                .setMaxWidthLimit(84));
+
+
+        return builder.build(getHolder(), entityPlayer);
+    }
+
+    @Override
+    public void addInformation(ItemStack stack, @Nullable World player, List<String> tooltip, boolean advanced) {
+        tooltip.add(I18n.format("gregtech.universal.tooltip.voltage_in", energyContainer.getInputVoltage(), GTValues.VN[getTier()]));
+        tooltip.add(I18n.format("gregtech.universal.tooltip.energy_storage_capacity", energyContainer.getEnergyCapacity()));
+        if (getTier() == 1)
+            tooltip.add(I18n.format("gregtech.machine.miner.description.lv", getType().radius, getType().tick / 20, GTValues.V[getTier() - 1]));
+        else if (getTier() == 2)
+            tooltip.add(I18n.format("gregtech.machine.miner.description.mv", getType().radius, getType().tick / 20, GTValues.V[getTier() - 1]));
+        else if (getTier() == 3)
+            tooltip.add(I18n.format("gregtech.machine.miner.description.hv", getType().radius, getType().tick / 20, GTValues.V[getTier() - 1]));
+    }
+
+    public boolean drainEnergy() {
+        if (energyContainer.getEnergyStored() >= energyPerTick && !done && !invFull) {
+            energyContainer.removeEnergy(energyPerTick);
+            return true;
+        }
+        return false;
+    }
+
+
+    @Override
+    public void update() {
+        super.update();
+        if (!getWorld().isRemote) {
+            if (!drainEnergy()) {
+                if (!done && x.get() == Integer.MAX_VALUE && y.get() == Integer.MAX_VALUE && z.get() == Integer.MAX_VALUE) {
+                    initPos();
+                }
+                if (invFull && getOffsetTimer() % 5 == 0) {
+                    pushItemsIntoNearbyHandlers(getFrontFacing());
+                    invFull = false;
+                }
+                return;
+            }
+
+            WorldServer world = (WorldServer) this.getWorld();
+
+
+            if (mineY.get() < tempY.get()) {
+                world.destroyBlock(new BlockPos(getPos().getX(), tempY.get(), getPos().getZ()), false);
+                tempY.decrementAndGet();
+                this.pipeY++;
+                writeCustomData(-200, b -> b.writeInt(pipeY));
+                markDirty();
+            }
+
+            if (blockPos.isEmpty()) {
+                blockPos = IMiner.getBlocksToMine(this, x, y, z, startX, startZ, startY, aRadius);
```

Comment:
I'm glad to see the optimization with LinkedList, but I don't know what the advantage of using LinkedList is? How do I feel that it is still retrieving all mines in one tick? It's not dynamic, or am I missing something?